federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+// the function returns any error caught, and the flag indicates if retryable
+func (s *ServiceController) processServiceForCluster(clusterName string, service *v1.Service, client *clientset.Clientset) error {
+	glog.V(4).Infof("process service %s/%s for cluster %s", service.Namespace, service.Name, clusterName)
+	// Create or Update k8s Service
+	err := s.ensureClusterServices(clusterName, service, client)
+	if err != nil {
+		glog.V(4).Infof("Failed to process service %s/%s for cluster %s", service.Namespace, service.Name, clusterName)
+		return err
+	}
+	glog.V(4).Infof("Successfully process service %s/%s for cluster %s", service.Namespace, service.Name, clusterName)
+	return nil
+}
+
+// Returns whatever error occurred along with a boolean indicator of whether it
+// should be retried.
+func (s *ServiceController) updateFederationService(key string, fedService *v1.Service) (error, bool) {
+	// Clone federation service, and create them in underlying k8s cluster
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return err, !retryable
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return fmt.Errorf("Unexpected service cast error : %v\n", service), !retryable
+	}
+
+	// handle available clusters one by one
+	var hasErr bool
+	for clusterName, cluster := range s.clusterCache.clientMap {
+		go func(){
+			err = s.processServiceForCluster(clusterName, service, cluster.clientset)
+			if err != nil {
+				hasErr = true
+			}
+		}()
+	}
+	if hasErr {
+		// detail error has been dumpped inside the loop
+		return fmt.Errorf("service %s/%s was not successfully updated to all clusters", service.Namespace, service.Name), retryable
+	}
+	return nil, !retryable
+}
+
+func (s *ServiceController) deleteFederationService(service *v1.Service) (error, bool) {
+	// handle available clusters one by one
+	for clusterName, cluster := range s.clusterCache.clientMap {
+		s.deleteClusterService(clusterName, service, cluster.clientset)
+	}
+	return nil, !retryable
+}
+func (s *ServiceController) deleteClusterService(clusterName string, service *v1.Service, clientset *clientset.Clientset) (error, bool) {
+	glog.V(4).Infof("Deleting service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		err = clientset.Core().Services(service.Namespace).Delete(service.Name, &api.DeleteOptions{})
+		if err == nil || errors.IsNotFound(err) {
+			glog.V(4).Infof("service %s/%s deleted from cluster %s", service.Namespace, service.Name, clusterName)
+			return nil, !retryable
+		}
+		time.Sleep(clientRetryInterval)
+	}
+	glog.V(4).Infof("failed to delete service %s/%s from cluster %s, %+v", service.Namespace, service.Name, clusterName, err)
+	return err, !retryable
+}
+
+func (s *ServiceController) ensureClusterServices(clusterName string, service *v1.Service, client *clientset.Clientset) error {
```

Comment:
This is only for one service in one cluster, right?  Then rather rename to ensureClusterService (singular).